### PR TITLE
Add 2018-specific changes.

### LINKED ISSUE #91: Include a page about 2018-only changes
Include a dedicated page that explains exactly what parts that are only available on 2018 + all the breakage.
This should be a tiny list.

### PR DESCRIPTION
**Pull request by ehuss**
Closes #91.

### CODE REVIEW

**Comment by ehuss** [comment]
Some notes:
- I have no idea where to place the page. I just stuck it at the end. Another option would be to place it directly in `rust-2018/index.md` but that didn't quite feel right.
- This is intentionally terse since it is intended to just be an overview. I'd be happy to add more detail, examples, links, etc. if you want.
- I'd be fine with dropping the Cargo section. The changes are relatively minor, and Cargo gives plenty of warnings in both editions.
- Feel free to suggest anything. This is the information *I* want to see, but I understand that the style/voice differs from the rest of the guide.
- I'll try to keep this list updated in the future. Some of these are actively being moved to 2015, and things like async/await will likely need to be added.

File: src/rust-2018/edition-changes.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# 2018-Specific Changes
+
+The following is a summary of changes that only apply to code compiled with
+the 2018 edition compared to the 2015 edition.
+
+## Rust
+- [Non-lexical lifetimes].
```
Comment (Centril) [inline src/rust-2018/edition-changes.md]: Might want to note that this is a temporary arrangement; we intend to backport later.

File: src/rust-2018/edition-changes.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+- [At most once] `?` macro repetition operator.
+- [Path changes]:
+    - `use` declarations must begin with:
+        - `crate` – refers to the current crate.
+        - `self` – refers to the current module.
+        - `super` – refers to the parent module.
+        - An external crate name.
+        - `::` – must be followed by an external crate name. This is required
+          if an external crate has the same name as an in-scope item, to catch
+          possible ambiguities with future module changes.
```
Comment (Centril) [inline src/rust-2018/edition-changes.md]: We'll introduce uniform_paths eventually; might want to hint at that as well.

File: src/SUMMARY.md [hunk 3]
```diff
@@ -83,4 +83,5 @@
       - [Global allocators](rust-2018/platform-and-target-support/global-allocators.md)
       - [MSVC toolchain support](rust-2018/platform-and-target-support/msvc-toolchain-support.md)
       - [MUSL support for fully static binaries](rust-2018/platform-and-target-support/musl-support-for-fully-static-binaries.md)
-      - [`cdylib` crates for C interoperability](rust-2018/platform-and-target-support/cdylib-crates-for-c-interoperability.md)
\ No newline at end of file
+      - [`cdylib` crates for C interoperability](rust-2018/platform-and-target-support/cdylib-crates-for-c-interoperability.md)
+    - [2018-Specific Changes](rust-2018/edition-changes.md)
```
Comment (Centril) [inline src/SUMMARY.md]: Might want to move this to be the first item, i.e. 3.1 (and modules become 3.2).

File: src/rust-2018/edition-changes.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    - Paths in `pub(in path)` visibility modifiers must start with `crate`,
+      `self`, or `super`.
+- [Anonymous trait function parameters] are not allowed.
+    - [Trait function parameters] may use any irrefutable pattern when the
+      function has a body.
+- [`dyn`] is a [strict keyword], in 2015 it is a [weak keyword].
+- `async`, `await`, and `try` are [reserved keywords].
```
Comment (petrochenkov) [inline src/rust-2018/edition-changes.md]: `await` is not reserved

Comment (Centril) [reply src/rust-2018/edition-changes.md]:
```rust
error[E0721]: `await` is a keyword in the 2018 edition
 --> src/lib.rs:1:4
  |
1 | fn await() {}
  |    ^^^^^ help: you can use a raw identifier to stay compatible: `r#await`
```

You have to use `#![feature(await_macro)]` to make it unreserved.

Comment (petrochenkov) [reply src/rust-2018/edition-changes.md]: Oh god.
They future-proofed it with a hack in a lint pass.

Ok, this makes it kinda reserved in practice.

**Comment by ehuss** [comment]
Good points, thanks for taking a look!

**Comment by dwijnand** [comment]
This would be lovely to land, it took me quite some digging before I realised that NLL is edition 2018 only.

As a side note, I couldn't figure out how to specify the edition to `rustc` because it's not in `rustc --help` or in the guide anywhere, so I ended up looking it up in the cargo sources.  (I later found it in `rustc --help -v`.)

**Comment by steveklabnik** [comment]
Thank you! People have been asking about this for a while. I think this is a fine place to put it, we can iterate if it turns out to be not ideal in practice.

**Comment by dwijnand** [comment]
Now that Rust nightly is fixed, could someone rerun the merge build for this so it deploys, please?  Thank you!

https://travis-ci.com/rust-lang-nursery/edition-guide/builds/96925518

**Comment by ehuss** [comment]
I poked it, but it looks like it will be queued for a long time due to other jobs running.

**Comment by dwijnand** [comment]
https://rust-lang-nursery.github.io/edition-guide/rust-2018/edition-changes.html :tada:

**Pull request merged**